### PR TITLE
Swift Keylogger Mettle extension

### PR DESCRIPTION
**Pull request by pbarry-r7**
Associated [MSF PR is 9635](https://github.com/rapid7/metasploit-framework/pull/9635).

We'll likely want to update the `-target x86_64-apple-macosx10.13` build option in the mettle/extensions/keylogger/Makefile.am file to support 10.9, which I believe is the earliest macOS version this version of Swift can support.

## Verification Steps

See [MSF PR 9635](https://github.com/rapid7/metasploit-framework/pull/9635)

### CODE REVIEW

File: mettle/extensions/keylogger/module.modulemap [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module swift2mettle [system] {
+	header "/Users/pebarry/mettle/mettle/extensions/keylogger/swift2mettle.h"
```
Comment (timwr) [inline mettle/extensions/keylogger/module.modulemap]: header "swift2mettle.h" :trollface:

File: mettle/extensions/keylogger/Keylogger.swift [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+//  Keylogger
+//
+//  Created by Skrew Everything on 14/01/17.
+//  Copyright © 2017 Skrew Everything. All rights reserved.
+//
+
+import Foundation
+import IOKit.hid
+import Cocoa
+import let swift2mettle.KEYLOGGER_DATA_DIR
```
Comment (timwr) [inline mettle/extensions/keylogger/Keylogger.swift]: I'm not sure what github is doing here with the highlighting

**Comment by bwatters-r7** [comment]
On hold until we get the builder fixed in our environment.

**Comment by timwr** [comment]
https://github.com/pbarry-r7/mettle/pull/3

**Comment by jmartin-r7** [comment]
I originally read the 10.9 requirement to mean we needed 10.9 to build this, turns out swift was not provided on in xcode until 10.10 so will test build in 10.10 and target 10.9.

**Comment by jmartin-r7** [comment]
Attempted to build this code against older swift toolchain with the original `Makefile.am` suggestion to target 10.9:
```
osx-10_10:mettle vagrant$ swiftc -v
Apple Swift version 2.1.1 (swiftlang-700.1.101.15 clang-700.1.81)
Target: x86_64-apple-darwin14.5.0
```

This shows that is uses syntax and apis that are part of Swift 3.0 available in [MacOS 10.11.5 and greater.](https://swift.org/download/#using-downloads)

```
Building mettle for native
Apple Swift version 2.1.1 (swiftlang-700.1.101.15 clang-700.1.81)
Target: x86_64-apple-macosx10.9
/Library/Developer/CommandLineTools/usr/bin/swift -frontend -c -primary-file /Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift /Users/vagrant/mettle/mettle/extensions/keylogger/Keylogger.swift /Users/vagrant/mettle/mettle/extensions/keylogger/main.swift -target x86_64-apple-macosx10.9 -enable-objc-interop -I /Users/vagrant/mettle/mettle/extensions/keylogger -color-diagnostics -module-name keylogger -o /var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/CallBackFunctions-3e40fb.o
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:29:25: error: expected member name following '.'
        if !FileManager.default.fileExists(atPath: path.path)
                        ^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:29:25: error: 'default' label can only appear inside a 'switch' statement
        if !FileManager.default.fileExists(atPath: path.path)
                        ^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:65:29: error: expected member name following '.'
            if !FileManager.default.fileExists(atPath: path.path)
                            ^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:65:29: error: 'default' label can only appear inside a 'switch' statement
            if !FileManager.default.fileExists(atPath: path.path)
                            ^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:113:25: error: expected member name following '.'
        if !FileManager.default.fileExists(atPath: path.path)
                        ^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:113:25: error: 'default' label can only appear inside a 'switch' statement
        if !FileManager.default.fileExists(atPath: path.path)
                        ^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:178:1: error: expected '}' at end of closure

^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:20:69: note: to match this opening '{'
    static let Handle_DeviceMatchingCallback: IOHIDDeviceCallback = { context, result, sender, device in
                                                                    ^
/Users/vagrant/mettle/mettle/extensions/keylogger/CallBackFunctions.swift:178:1: error: expected declaration

^
<unknown>:0: error: cannot load underlying module for 'CoreGraphics'
<unknown>:0: note: did you forget to set an SDK using -sdk or SDKROOT?
<unknown>:0: note: use "xcrun -sdk macosx swiftc" to select the default OS X SDK installed with Xcode
```

This leads to the question of what and how we want to do support of extensions that require specific environment tools chains and produces object what may have target OS dependencies?

**Comment by timwr** [comment]
I think it would be reasonable to only target the latest version if it's too much effort to re-write entirely for older versions.
That said you raise a good point about how to target specific OS dependencies. I'm not sure of the best way, but appending the version to build triple seems like a good move. We could also considering renaming the extension itself, e.g `keylogger-10.9`

**Comment by jmartin-r7** [comment]
We have decided to update our build infrastructure to baseline at macOS 10.11.

The swift code here is not compatible with 10.11 however looks like it may be possible to back port.

**Comment by jmartin-r7** [comment]
Closing as a good pattern for including/compiling plugins that have varying iOS api level requirements in automation has not been established.

**Pull request closed without merging**